File: examples/guests/lua/reporter/reporter.lua
local ffi = require('ffi')
local polyplug = require('polyplug_guest')
local polyplug_abi = require('polyplug_abi')
local contracts = require('generated.guest.contracts')

-- Factory returning a per-instance reporter. The reporter is stateless, so each
-- instance is a fresh object whose method is the contract function.
local function new_reporter(host)
    local self = {}

    function self:report(input)
        local s = polyplug_abi.to_str(input)
        if s:sub(1, 12) == 'TRANSFORMED:' then s = s:sub(13) end
        local name, value, count = s:match('^([^|]*)|([^|]*)|([^|]*)$')
        if name then
            return polyplug.alloc_string(string.format(
                'Report: %s has value \'%s\' with count %s', name, value, count))
        end
        return polyplug.alloc_string('INVALID:format')
    end

    return self
end

contracts.set_reporter_factory(new_reporter)

return contracts

File: examples/guests/csharp/reporter/Reporter.cs
using System.Runtime.CompilerServices;
using Polyplug.Guest;
using Polyplug.Abi;

namespace Reporter;

public sealed class ReporterPlugin : IDataReporterGuestContract
{
    // Host handle for this runtime, captured at instance creation.
    private readonly IntPtr _host;

    public ReporterPlugin(IntPtr host)
    {
        _host = host;
    }

    public StringView Report(StringView input)
    {
        string s = StringViewHelper.StripPrefix(input, "TRANSFORMED:");
        string[] parts = s.Split('|');
        if (parts.Length >= 3)
        {
            return PolyplugHost.AllocString(_host, $"Report: {parts[0]} has value '{parts[1]}' with count {parts[2]}");
        }
        return PolyplugHost.AllocString(_host, "INVALID:format");
    }
}

public static class Registration
{
    [ModuleInitializer]
    public static void Register()
    {
        // The factory receives the HostApi pointer per created instance.
        ReporterInterfaces.SetReporterFactory(host => new ReporterPlugin(host));
    }
}
